tfhe/docs/tutorial/sha256_bool.md

```diff
@@ -0,0 +1,332 @@
+        for i in 0..64 {
+            let temp1 = add(add(add(add(h, ch(&e, &f, &g)), w[i]), hex_to_bools(K[i])), sigma_upper_case_1(&e));
+            let temp2 = add(sigma_upper_case_0(&a), maj(&a, &b, &c));
+            h = g;
+            g = f;
+            f = e;
+            e = add(d, temp1);
+            d = c;
+            c = b;
+            b = a;
+            a = add(temp1, temp2);
+        }
+
+        hash[0] = add(hash[0], a);
+        hash[1] = add(hash[1], b);
+        hash[2] = add(hash[2], c);
+        hash[3] = add(hash[3], d);
+        hash[4] = add(hash[4], e);
+        hash[5] = add(hash[5], f);
+        hash[6] = add(hash[6], g);
+        hash[7] = add(hash[7], h);
+    }
+
+    // Concatenate the final hash values to produce a 256-bit hash
+    let mut output = [false; 256];
+    for i in 0..8 {
+        output[i * 32..(i + 1) * 32].copy_from_slice(&hash[i]);
+    }
+    output
+}
+```
+
+## Making it homomorphic
+
+The key idea is that we can replace each bit of ```padded_input``` with a Fully Homomorphic Encryption of the same bit value, and operate over the encrypted values using homomorphic operations. To achieve this we need to change the function signatures and deal with the borrowing rules of the Ciphertext type (which represents an encrypted bit) but the structure of the sha256 function remains the same. The part of the code that requires more consideration is the implementation of the sha256 operations, since they will use homomorphic boolean operations internally.
+
+Homomorphic operations are really expensive, so we have to remove their unnecessary use and maximize parallelization in order to speed up the program. To simplify our code we use the Rayon crate which provides parallel iterators and efficiently manages threads. Let's now take a look at each sha256 operation!
+
+#### Rotate Right and Shift Right
+
+As we have highlighted, these two operations can be evaluated by changing the position of each encrypted bit in the word, thereby requiring 0 homomorphic operations. Here is our implementation:
+
+```rust
+fn rotate_right(x: &[Ciphertext; 32], n: usize, sk: &ServerKey) -> [Ciphertext; 32] {
+    let mut result = trivial_bools(&[false; 32], sk);
+    for i in 0..32 {
+        result[(i + n) % 32] = x[i].clone();
+    }
+    result
+}
+
+fn shift_right(x: &[Ciphertext; 32], n: usize, sk: &ServerKey) -> [Ciphertext; 32] {
+    let mut result = trivial_bools(&[false; 32], sk);
+    for i in 0..(32 - n) {
+        result[i + n] = x[i].clone();
+    }
+    result
+}
+```
+We see a function called ``trivial_bools`` that we use along our implementation to initialize an array of 32 Ciphertexts. This is because we cannot copy the same Ciphertext for each position of the array as we do with simple bools, so we create 32 different trivial encryptions. We will also use this function in order to trivially encrypt constant values to operate homomorphically with them.
+
```

Comment:
not sure I understand here, you could clone x to be the mutable result and the call rotate/shift ops from rust right ?

https://doc.rust-lang.org/std/primitive.slice.html#method.rotate_right for the shift may need to do some manual management at the end

I guess the code could be updated as well then :)

Reply:
That's a good point. About the highlighted text, I meant that we cannot initialize arrays of ciphertexts like this:
```let a = [sk.trivial_encrypt(false); 32];```

Reply:
ok yes